invenio_oauthclient/handlers/utils.py

```diff
@@ -113,39 +113,44 @@ def create_or_update_roles(groups):
-            if existing_role_by_name and existing_role_by_name.is_managed:
-                current_app.logger.exception(
-                    f'Error while syncing roles: A managed role with name: ${group["name"]} already exists'
-                )
-                continue
-            if not existing_role:
-                role = current_datastore.create_role(
-                    id=group["id"],
-                    name=group["name"],
-                    description=group.get("description"),
-                    is_managed=False,
-                )
-                roles_ids.add(role.id)
-            elif existing_role and _role_needs_update(existing_role, group):
-                role_to_update = Role(
-                    id=group["id"],
-                    name=group["name"],
-                    description=group.get("description"),
-                    is_managed=False,
-                )
-                role = current_datastore.update_role(role_to_update)
-                roles_ids.add(role.id)
-            else:
-                roles_ids.add(existing_role.id)
-
+            with db.session.begin_nested():
+                current_app.logger.debug(f"Syncing role: {group['name']}")
+                
+                existing_role = current_datastore.find_role_by_id(group["id"])
+                if existing_role and existing_role.is_managed:
+                    current_app.logger.exception(
+                        f'Error while syncing roles: A managed role with id: {group["id"]} already exists'
+                    )
+                    continue
+                
+                existing_role_by_name = current_datastore.find_role(group["name"])
+                if existing_role_by_name and existing_role_by_name.is_managed:
+                    current_app.logger.exception(
+                        f'Error while syncing roles: A managed role with name: {group["name"]} already exists'
+                    )
+                    continue
+
+                if not existing_role:
+                    role = current_datastore.create_role(
+                        id=group["id"],
+                        name=group["name"],
+                        description=group.get("description"),
+                        is_managed=False,
+                    )
+                    roles_ids.add(role.id)
+                elif existing_role and _role_needs_update(existing_role, group):
+                    role_to_update = Role(
+                        id=group["id"],
+                        name=group["name"],
+                        description=group.get("description"),
+                        is_managed=False,
+                    )
+                    role = current_datastore.update_role(role_to_update)
+                    roles_ids.add(role.id)
+                else:
+                    roles_ids.add(existing_role.id)
+                
+                db.session.flush()  # Ensure changes are written before committing
```

Comment:
Isn't that happening eitherway on `exit` of `db.session.begin_nested` ?

Reply:
[yes](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html):

> When the context manager yielded by [Session.begin_nested()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.begin_nested) completes, it “commits” the savepoint, which includes the usual behavior of flushing all pending state. When an error is raised, the savepoint is rolled back and the state of the [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) local to the objects that were changed is expired.

I am wondering if we should delegate to the caller to call this method inside the `begin_nested`, WDYT?

Reply:
I thought about this, however kept it here as it was the same behaviour as before. I will change it to be the caller 👍